contracts/extensions/ReputationBootstrapper.sol

```diff
@@ -0,0 +1,148 @@
+
+import "./../../lib/dappsys/erc20.sol";
+import "./ColonyExtensionMeta.sol";
+
+// ignore-file-swc-108
+
+
+contract ReputationBootstrapper is ColonyExtensionMeta {
+
+  // Constants
+
+  uint256 constant INT128_MAX = 2**127 - 1;
+  // Based on a 90-day half-life
+  uint256 constant HOURLY_DECAY_NUMERATOR = 999679150010889; // e ** -(ln(2)/2160)
+  uint256 constant DAILY_DECAY_NUMERATOR =  992327946262943; // e ** -(ln(2)/90)
+  uint256 constant DECAY_DENOMINATOR =     1000000000000000;
+
+  // Events
+
+  event GrantSet(bytes32 hashedSecret, uint256 reputationAmount);
+  event GrantClaimed(address recipient, uint256 reputationAmount, uint256 tokenAmount);
+
+  // Data structures
+
+  struct Grant {
+    uint256 amount;
+    uint256 timestamp;
+  }
+
+  // Storage
+
+  address public token;
+  bool public giveTokens;
+
+  mapping (bytes32 => Grant) public grants;
+
+  // Modifiers
+
+  modifier onlyRoot() {
+    require(colony.hasUserRole(msgSender(), 1, ColonyDataTypes.ColonyRole.Root), "reputation-bootsrapper-caller-not-root");
+    _;
+  }
+
+  // Overrides
+
+  /// @notice Returns the identifier of the extension
+  function identifier() public override pure returns (bytes32) {
+    return keccak256("ReputationBootstrapper");
+  }
+
+  /// @notice Returns the version of the extension
+  function version() public override pure returns (uint256) {
+    return 1;
+  }
+
+  /// @notice Configures the extension
+  /// @param _colony The colony in which the extension holds permissions
+  function install(address _colony) public override auth {
+    require(address(colony) == address(0x0), "extension-already-installed");
+
+    colony = IColony(_colony);
+    token = colony.getToken();
+  }
+
+  /// @notice Called when upgrading the extension
+  function finishUpgrade() public override auth {}
+
+  /// @notice Called when deprecating (or undeprecating) the extension
+  function deprecate(bool _deprecated) public override auth {
+    deprecated = _deprecated;
+  }
+
+  /// @notice Called when uninstalling the extension
+  function uninstall() public override auth {
+    uint256 balance = ERC20(token).balanceOf(address(this));
+    require(ERC20(token).transfer(address(colony), balance), "reputation-bootstrapper-transfer-failed");
+
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  // Public
+
+  function setGiveTokens(bool _giveTokens) public onlyRoot {
+    giveTokens = _giveTokens;
+  }
+
+  function setGrants(bytes32[] memory _hashedSecrets, uint256[] memory _amounts) public onlyRoot notDeprecated {
+    require(_hashedSecrets.length == _amounts.length, "reputation-bootsrapper-invalid-arguments");
+
+    for (uint256 i; i < _hashedSecrets.length; i++) {
+      require(_amounts[i] <= INT128_MAX, "reputation-bootstrapper-invalid-amount");
+      grants[_hashedSecrets[i]] = Grant(_amounts[i], block.timestamp);
+
+      emit GrantSet(_hashedSecrets[i], _amounts[i]);
+    }
+  }
+
+  function claimGrant(uint256 _secret) public {
```

Comment:
Unfortunately, I think this type of claim process is fundamentally flawed. If I see a pending transaction for `claimGrant`, I can front-run the transaction with the secret from my own account with a higher gas price, and 'snipe' it.

Reply:
I see that. One solution which comes to mind would be to require claiming addresses to have a `joincolony.eth` ENS name before the claim. That would require users to go through the onboarding flow before presenting their secret, and would help alleviate this risk (but not remove it entirely). What do you think of that? I suppose miners could go ahead and create their ENS names before proceeding to run this attack.

Another idea would be to require a two-part "handshake" where users present the secret and has their address recorded, but ultimately the admin has to go back and hit the button to actually transfer the funds. That would be a product decision though, so if you think it's promising we should rope in the rest of the crew.

Reply:
> I suppose miners [sic] could go ahead and create their ENS names before proceeding to run this attack.

Indeed.

>Another thing we could say is that this contract is knowingly vulnerable to this attack, and so should be used only for small amounts of bootstrapping funds. I expect you won't like that solution, but we can put it on the table.

I do indeed not like that solution!

> ultimately the admin has to go back and hit the button to actually transfer the funds.

There seems very little benefit to using this extension, in that case?

I think the only possible on-chain solution is a commit-and-reveal on the user's side as well, with a minimum timestamp between those two transactions to prevent front-running of both transactions.

I do think product needs to be involved to sign off on any of this though.

Reply:
@arrenv what are your thoughts on this? Alex is saying that the design as-is is vulnerable to miner front-running. One solution would be to require two transactions, the first essentially "registering" the address and the intended hashed secret, and the second providing the secret itself, checking that the caller is the initial address. By introducing a delay (on the order of ~30 seconds or so I would suppose, maybe less) then a miner could not front-run. Would this two-transactions-plus-delay flow be acceptable?

Reply:
I would want it to be at least an hour

Reply:
That seems like a very long amount of time and meaningfully affects the UX. I would expect an experience along the lines of a token bridge (e.g. two transactions, one on each chain, 15 or so blocks apart) would be more palatable to users.

Reply:
How long have you had to wait for a transaction to be mined in the past? If the timescale is smaller than that, even a commit-and-reveal can be front-run.

Reply:
When bridging from xdai to mainnet it's quick, but that's because we're waiting on the xdai blocks which are snappy. If we're waiting on mainnet then I suppose it would be longer, although post-merge maybe not so much. In any case, what do you think of phrasing the delay in terms of block confirmations and not an absolute amount of time?

Reply:
Having the two transactions sounds acceptable to me, then using a block confirmation as a delay. As it would then be compatible on different chains. In the UI we can then use an approx. time for making the claim based on the chains block rate.

Reply:
Why is block confirmations better than an absolute amount of time?

Reply:
I would say because block confirmations get more directly at the issue we are about, which is preventing miners from mining transactions to benefit themselves. The more blocks are confirmed, the harder it becomes for a miner to selfishly mine. Further, block confirmations allow us to move more quickly on chains with short block times, while still providing security on slower chains, waiting "an hour" if need be. If we waited a fixed amount of time, that could either be way too long (on fast chains), or potentially still not long enough (if for whatever reason transactions take a very long time to mine).

Reply:
> more directly at the issue we are about, which is preventing miners from mining transactions to benefit themselves

This isn't really the issue, nor is miners rewriting history as you seem to be implying. The issue is:

* A user commits `keccak(walletAddress, secret)` (transaction 1)
* That user then reveals `secret` via another transaction (transaction 2), claiming their reputation 
* A malicious user sees transaction2 in the pending transaction pool, and commits `keccak(walletAddress2, secret` and then reveals `secret`, all before transaction 2 is mined (because they're paying very high gas fees).

The 'honest' transactions are likely to take a little while to be mined, as they are almost certainly metatransactions that we're paying for, and so going to be at the low end of the fee scale. The likely time for those transactions to be mined has to be _significantly_ lower than the minimum allowed time between a commit and reveal.

Reply:
I'm following. The issue at hand is less the amount of blocks since the initial tx was confirmed and more about making sure that a slower tx isn't scooped up by a faster one.

Sounds like the timeline you have in mind is:

```
t0         [... hour... ]          t1                                     [...hour...]              t2 
secret committed           secret revealed
                                         malicious user commits secret                   malicious user reveals secret
```

You are concerned with making sure that the honest transaction at t1 is mined before the dishonest one at t2.